src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/UseNameofInPlaceOfString.cs

```diff
@@ -0,0 +1,141 @@
+                                                                         customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleWithSuggestion);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(AnalyzeArgument, OperationKind.Argument);
+        }
+
+        private void AnalyzeArgument(OperationAnalysisContext context)
+        {
+
+            var argument = (IArgumentOperation)context.Operation;
+
+            if (argument.Value.Kind != OperationKind.Literal)
+            {
+                return;
+            }
+            // TODO better way to get the string?
+            var stringText = argument.Value.ConstantValue.Value.ToString();
+            var properties = ImmutableDictionary<string, string>.Empty.Add("StringText", stringText);
+
+            // TODO when showing diagnostic on a named argument, should just squiggle the string and not the argument name
+            var matchingParameter = (IParameterSymbol)argument.Parameter;
+            switch (matchingParameter.Name)
+            {
+                case ParamName:
+                    var parametersInScope = GetParametersInScope(context);
+                    // TODO if argument doesn't match any parameters, give a warning
+                    if (HasAMatchInScope(stringText, parametersInScope))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(RuleWithSuggestion, argument.Syntax.GetLocation(), properties: properties));
+                    }
+                    return;
+                case PropertyName:
+                    var propertiesInScope = GetPropertiesInScope(context);
+                    if (HasAMatchInScope(stringText, propertiesInScope))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(RuleWithSuggestion, argument.Syntax.GetLocation(), properties: properties));
+                    }
+                    return;
+                default:
+                    return;
+            }
+        }
+
+        private IEnumerable<string> GetPropertiesInScope(OperationAnalysisContext context)
+        {
+            var containingType = context.ContainingSymbol.ContainingType;
+            if (containingType != null)
+            {
+                foreach (var property in containingType.GetMembers().OfType<IPropertySymbol>())
+                {
+                    yield return property.Name;
+                }
+            }
+        }
+
+        internal IEnumerable<string> GetParametersInScope(OperationAnalysisContext context)
+        {
+            foreach (var parameter in context.ContainingSymbol.GetParameters())
```

Comment:
Can you please add a comment explaining the logic here and the below while loop?